libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,560 @@
+			return nullopt;
+		},
+		[&](CFG::BasicBlock::ConditionalJump const& _conditionalJump) -> std::optional<Stack>
+		{
+			bool zeroVisited = _blocksWithExitLayout.count(_conditionalJump.zero);
+			bool nonZeroVisited = _blocksWithExitLayout.count(_conditionalJump.nonZero);
+			if (zeroVisited && nonZeroVisited)
+			{
+				Stack stack = combineStack(
+					m_layout.blockInfos.at(_conditionalJump.zero).entryLayout,
+					m_layout.blockInfos.at(_conditionalJump.nonZero).entryLayout
+				);
+				stack.emplace_back(_conditionalJump.condition);
+				return stack;
+			}
+			if (!zeroVisited)
+				_toVisit.emplace_front(_conditionalJump.zero);
+			if (!nonZeroVisited)
+				_toVisit.emplace_front(_conditionalJump.nonZero);
+			return nullopt;
+		},
+		[&](CFG::BasicBlock::FunctionReturn const& _functionReturn) -> std::optional<Stack>
+		{
+			yulAssert(_functionReturn.info, "");
+			Stack stack = _functionReturn.info->returnVariables | ranges::views::transform([](auto const& _varSlot){
+				return StackSlot{_varSlot};
+			}) | ranges::to<Stack>;
+			stack.emplace_back(FunctionReturnLabelSlot{});
+			return stack;
+		},
+		[&](CFG::BasicBlock::Terminated const&) -> std::optional<Stack>
+		{
+			return Stack{};
+		},
+	}, _block.exit);
+}
+
+void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const& _entry)
+{
+	list<CFG::BasicBlock const*> toVisit{&_entry};
+	set<CFG::BasicBlock const*> visited;
+
+	// TODO: check whether visiting only a subset of these in the outer iteration below is enough.
+	list<pair<CFG::BasicBlock const*, CFG::BasicBlock const*>> backwardsJumps = collectBackwardsJumps(_entry);
+
+	while (!toVisit.empty())
+	{
+		// First calculate stack layouts without walking backwards jumps, i.e. assuming the current preliminary
+		// entry layout of the backwards jump target as the initial exit layout of the backwards-jumping block.
+		while (!toVisit.empty())
+		{
+			CFG::BasicBlock const *block = *toVisit.begin();
+			toVisit.pop_front();
+
+			if (visited.count(block))
+				continue;
+
+			if (std::optional<Stack> exitLayout = getExitLayoutOrStageDependencies(*block, visited, toVisit))
+			{
+				visited.emplace(block);
+				// We can skip the visit, if we have seen this precise exit layout already last time.
+				// Note: if the entire graph is revisited in the backwards jump check below, doing
+				//       this seems to break things; not sure why.
+				// Note: since I don't quite understand why doing this can break things, I comment
+				//       it out for now, since not aborting in those cases should always be safe.
+				// if (auto* previousInfo = util::valueOrNullptr(m_layout.blockInfos, block))
+				//	if (previousInfo->exitLayout == *exitLayout)
+				//		continue;
+				auto& info = m_layout.blockInfos[block];
+				info.exitLayout = *exitLayout;
+				info.entryLayout = propagateStackThroughBlock(info.exitLayout, *block);
+
+				for (auto entry: block->entries)
+					toVisit.emplace_back(entry);
+			}
+			else
+				continue;
```

Comment:
Why don't we need to re-add `block` to the end of `toVisit`?

Reply:
Oh because of line 363

Reply:
Yeah, that's maybe also a bit convoluted...
This case only happens if we didn't get an exit layout, in which case the exit blocks are added to ``toVisit``... eventually I arrive at some exit, which I can visit, and in the end of the "true" case above, that will then add its *entries* to ``toVisit``, so we'll eventually get back to ``block`` that way...
Which makes sense, since only after visiting an exit, it actually makes sense to revisit the entries.